src/Material/Menu.elm

```diff
@@ -0,0 +1,523 @@
+  , enabled = enabled
+  , divider = divider
+  }
+
+
+-- ACTION, UPDATE
+
+
+{-| Component action. The `Select n` action fires when the `n`th Item is selected; `Open`, `Close` when the menu opens, closes.
+-}
+
+-- Note: The Close' index is 1-based while the Select index is 0-based.
+
+type Action =
+    Toggle Oracle
+  | Close' Int Oracle
+  | Hide Int Oracle
+
+  | Select Int
+  | Open
+  | Close
+
+  | Ripple Int Ripple.Action
+
+
+{-| Component update.
+-}
+update : Action -> Model -> (Model, Effects Action)
+update action model =
+
+  case action of
+
+
+    -- If the menu is closed, we transition into opening state.
+    Toggle oracle -> effect
+
+        (Effects.tick (\_ -> Open))
+
+        { model | open =
+                    case model.open of
+                      Nothing   -> Just False
+                      Just True -> Nothing
+                      _         -> model.open
+
+                , oracle =
+                    case model.open of
+                      Nothing   -> Just oracle
+                      Just True -> Just oracle
+                      _         -> model.oracle
+
+
+                , items =
+                    [1..List.length oracle.offsetTops]
+                    |> List.map (\i -> (i, Ripple.model))
+                    |> Dict.fromList
+        }
+
+    -- Transition from opening state to open state.
+    Open -> pure
+      { model | open = if model.open == Just False then
+                           Just True
+                         else
+                           model.open
+      }
+
+    -- Immediately close the menu.
+    Hide idx oracle -> effect
+
+      ( Effects.tick (\_ -> Select idx) )
+
+      { model | open = Nothing, closing = False, oracle = Just oracle }
+
+    -- Close the menu after some delay.
+    Close' idx oracle -> effect
+
+      ( Effects.task
+        <| Task.andThen (Task.sleep constant.closeTimeout) << always
+        <| Task.succeed (Hide idx oracle))
+
+      { model | closing = True }
+
+
+    Ripple idx action ->
+
+      let
+
+        (model', effects) = Dict.get idx model.items
+                         |> Maybe.withDefault Ripple.model
+                         |> Ripple.update action
+      in
+
+        effect
+
+          ( Effects.map (Ripple idx) effects )
+
+          { model | items = Dict.insert idx model' model.items }
+
+    Select _ -> pure model
+    Close -> pure model
+
+
+-- VIEW
+
+
+{-| Component view.
+-}
+view : Signal.Address Action -> Model -> List Item -> List Html
+view addr model items =
+  [ styled button
+      [ cs "mdl-button"
+      , cs "mdl-js-button"
+      , cs "mdl-button--icon"
+      , Style.attribute (onClick addr Oracle.decode Toggle)
+      ]
+      [ Html.i [ Html.class "material-icons" ] [ Html.text "more_vert" ]
+      ]
+  , styled div
+      [ cs "mdl-menu__container"
+      , cs "is-upgraded"
+      , cs' "is-visible" (model.open == Just True)
+
+      , css' "width"
+          ( case model.oracle of
```

Comment:
This function contains `case model.oracle of Just oracle -> f oracle Nothing -> "auto" a lot. Couldn't the commonalities be split out into a separate functions?

(Elm follows Haskell enough that it is usually clearer to make short seemingly trivial helper functions rather than writing everything out inline, I think.)